Schemas.md

```diff
@@ -0,0 +1,521 @@
+## Table of Contents
+
+* [Introduction](#introduction)
+* [Write a schema](#write-a-schema)
+* [Uses for schemas in Oppia](#uses-for-schemas-in-oppia)
+  * [Schema-based forms](#schema-based-forms)
+  * [Validating controller arguments](#validating-controller-arguments)
+    * [Schema validation system code](#schema-validation-system-code)
+    * [How to write validation schema for handlers](#how-to-write-validation-schema-for-handlers)
+    * [Important code pointers](#important-code-pointers)
+      * [Default and optional arguments](#default-and-optional-arguments)
+      * [Domain object arguments: Schemas with type object_dict](#domain-object-arguments-schemas-with-type-object_dict)
+        * [Case 1: object_class key](#case-1-object_class-key)
+        * [Case 2: validation_method key](#case-2-validation_method-key)
+      * [Extra validators](#extra-validators)
+      * [Extra arguments](#extra-arguments)
+      * [Handlers with no arguments](#handlers-with-no-arguments)
+      * [Post schema operations](#post-schema-operations)
+    * [Common errors faced](#common-errors-faced)
+    * [Examples for reference](#examples-for-reference)
+    * [Debugging tricks](#debugging-tricks)
+    * [Contact](#contact)
+
+## Introduction
+
+At Oppia, we often want to describe the possible structure of some data. This data could be as simple as the boolean value of a checkbox in a user's preferences, or it could be more complex like the valid arguments to one of our backend API controllers. We use describe these data using _schemas_.
+
+Schemas describe the structure of data in a machine-readable format so that generic utility functions can operate on arbitrary data so long as the data comes with a schema. For example, schemas let us:
```

Comment:
Bit weird to refer to "schemas" at the start and "schema" at the end, in the first sentence. Maybe rephrase so that it flows more cleanly?

Reply:
done